### PR TITLE
pytorch 10.3 transformer

### PR DESCRIPTION
**Pull request by ypandya**
*Description of changes:*

By submitting this pull request, I confirm that you can use, modify,
copy, and redistribute this contribution, under the terms of your
choice.

### CODE REVIEW

File: chapter_preliminaries/ndarray.md [hunk 1]
```diff
@@ -55,7 +55,7 @@ To start, we import `tensorflow`. As the name is a little long, we often import
 it with a short alias `tf`.
 :end_tab:
 
-```{.python .input}
+```python
```
Comment (AnirudhDagar) [inline chapter_preliminaries/ndarray.md:58]: Why change this file? I guess this has been mistakenly committed?

**Comment by ypandya** [comment]
Thanks for rectifying that mistake. ndarray.md was not supposed to be part of the commit. One of the cells is timing out now. I'll check and see what I can do about this.

File: chapter_attention-mechanisms/transformer.md [hunk 2]
```diff
@@ -436,11 +738,34 @@ model = d2l.EncoderDecoder(encoder, decoder)
...
+src_vocab, tgt_vocab, train_iter = d2l.load_data_nmt(batch_size, num_steps)
+
+encoder = TransformerEncoder(
+    len(src_vocab), key_size, query_size, value_size, num_hiddens,
+    norm_shape, ffn_num_input, ffn_num_hiddens, num_heads,
+    num_layers, dropout)
+decoder = TransformerDecoder(
+    len(src_vocab), key_size, query_size, value_size, num_hiddens,
+    norm_shape, ffn_num_input, ffn_num_hiddens, num_heads,
+    dropout)
```
Comment (ypandya) [inline chapter_attention-mechanisms/transformer.md:758]:
```suggestion
    num_layers, dropout)
```

File: chapter_attention-mechanisms/transformer.md [hunk 3]
```diff
@@ -414,6 +687,35 @@ class TransformerDecoder(d2l.Decoder):
         return self.dense(X), state
 ```
 
+```{.python .input}
+#@tab pytorch
+class TransformerDecoder(d2l.Decoder):
+    def __init__(self, vocab_size, key_size, query_size, value_size,
+                 num_hiddens, norm_shape, ffn_num_input, ffn_num_hiddens,
+                 num_heads, dropout, **kwargs):
```
Comment (ypandya) [inline chapter_attention-mechanisms/transformer.md:695]:
```suggestion
                 num_heads, num_layers, dropout, **kwargs):
```

**Comment by ypandya** [comment]
The final training block for pytorch is taking 10x the time taken by the mxnet code. I've not been able to find the root cause this yet. There is not any significant differences in the implementation. The major difference in pytorch and mxnet implementation is that the input values for each linear layer needs to be provided explicitly for pytorch.

**Comment by AnirudhDagar** [comment]
@ypandya any updates on this?

**Comment by ypandya** [comment]
> @ypandya any updates on this?

I have matched the mxnet and pytorch models in detail. There is no difference in the architecture. My early guess was that there maybe difference in the layer norm implementation in pytorch and mxnet. But the training takes the same amount of time even when I experimented with replacing layernorm with batchnorm. I could not find any reason for the difference which can cause the torch training to take 10x the time taken by mxnet. It would be really helpful if someone with more knowledge of mxnet can take a look at this.

**Comment by AnirudhDagar** [comment]
Thanks! I see, I'll try to look at the PyTorch code if I can find the difference.
Maybe @astonzhang or @mli can have a look to spot the problem and to find a fix.

**Comment by StevenJokes** [comment]
cuda is slower than cpu??
image

**Comment by AnirudhDagar** [comment]
@ypandya I was able to resolve the issue and traced it back to a function defined in `seq2seq`  which needed some optimization. The implementation now is 60 fold faster than what it was before and it should pass CI as expected.

The whole problem was due to an inbuilt function which is present in MXNet and is missing in PyTorch. `sequence_mask` For now we have a good custom implementation for same, but I feel we can improve the code in this section in later revisions for a more fluent PyTorch Adaptation.

@astonzhang please take a look at the PR and review/merge. I think we should be good to go for now. Though I'll try to come up with a more elegant solution later.

**Review by AnirudhDagar — APPROVED** [review]

**Comment by mli** [comment]
Job d2l-en/PR-1361/5 is complete. 
Check the results at http://preview.d2l.ai/d2l-en/PR-1361/

**Comment by astonzhang** [comment]
Thanks @ypandya and @AnirudhDagar!
